src/components/navigation/navigation-item/navigation-item.component.ts

```diff
@@ -0,0 +1,133 @@
+    Renderer2,
+    ElementRef,
+    Optional,
+    SkipSelf,
+    Output,
+    AfterViewInit,
+    AfterContentInit,
+    OnDestroy
+} from '@angular/core';
+import { Router, NavigationEnd, UrlTree } from '@angular/router';
+import { filter } from 'rxjs/operators/filter';
+import { Subscription } from 'rxjs/Subscription';
+
+@Component({
+    selector: '[ux-navigation-item]',
+    templateUrl: './navigation-item.component.html',
+    host: {
+        '[class.active]': 'active',
+        '[class.selected]': 'expanded',
+    }
+})
+export class NavigationItemComponent implements AfterViewInit, AfterContentInit, OnDestroy {
+    @Input() header: string;
+    @Input() icon: string;
+    @Input() expanded: boolean = false;
+    @Input() link: string;
+
+    get active(): boolean {
+        if (this.link) {
+            return this._router.isActive(this.link, true);
+        }
+    }
+
+    level: number = 1;
+    indentWithoutArrow: boolean = true;
+
+    private _navigationEnd: Subscription;
+    private _childrenChanges: Subscription;
+
+    @ContentChildren(NavigationItemComponent, { descendants: true })
+    private _children: QueryList<NavigationItemComponent>;
+
+    get children(): NavigationItemComponent[] {
+        return this._children.filter(item => item !== this);
+    }    
+
+    constructor(
+        private _elementRef: ElementRef,
+        private _renderer: Renderer2,
+        @Optional()
+        @SkipSelf()
+        private _parent: NavigationItemComponent,
+        private _router: Router
+    ) {
+        this.level = _parent ? _parent.level + 1 : 1;
+
+        this._navigationEnd = _router.events.pipe(filter(event => event instanceof NavigationEnd)).subscribe(event => {              
+            // Set 'expanded' if link is active. If expanded, will set 'expanded' on ancestors.
+            this.hasActiveLink(this.link);
+        });
+    }
+
+    ngAfterViewInit(): void {
+        // Add classes to parent for styling
+        const parentListElement = this._elementRef.nativeElement.parentElement;
+        if (parentListElement) {
+            let levelClass: string = this.getLevelClass();
+            if (levelClass.length > 0) {
+                this._renderer.addClass(parentListElement, 'nav');
+                this._renderer.addClass(parentListElement, levelClass);
+            }
+        }
+    }
+
+    ngAfterContentInit(): void {
+        // Set 'indentWithoutArrow'
+        this.setIndentWithoutArrow();
+
+        // Update 'indentWithoutArrow' in response to changes to children
+        this._childrenChanges = this._children.changes.subscribe(() => this.setIndentWithoutArrow());
+    }
+
+    ngOnDestroy () {
+        this._navigationEnd.unsubscribe();
+        this._childrenChanges.unsubscribe();
+    }
+
+    private hasActiveLink(link: string | UrlTree): void {
+        if (link) {
+            this.expanded = this._router.isActive(link, true);            
+            if (this.expanded && this._parent) {
```

Comment:
This is still wrong. Revert to the original version and please address the review comment. Contact me for clarification if you do not understand.

Reply:
I have reverted to the original code but expansion of parents of a selected item does not occur, whether they themselves have a link or not. I have experimented with addition and removal of router links in the example HTML and with changes to the link definitions in the router module.

The call to hasActiveLink in the NavigationEnd handler sets the value of the 'expanded' flag according to the component's link. For parents of selected items this is always false. The recursive call within hasActiveLink will return 'true' but it doesn't set the expanded flag in the component.

What am I missing here?

Reply:
Something like this?

````
private hasActiveLink(link: string | UrlTree): boolean {
    if (link) {
        return this._router.isActive(link, true) || this.children.any((item) => item.hasActiveLink(item.link));
    }

    return false;
}
````

Explanation: Every component receives the NavigationEnd event. We use this to determine the state of the `expanded` property for each node, ensuring that the entire tree is always up to date with every navigation event. This avoids the need to go back and collapse siblings explicitly. This makes use of recursion, which is okay since the maximum depth is 5, and keeps the code tidy.

Let me know if there is anything I've overlooked with this, or anything that you don't understand. It's not a problem to ask for help, as previously agreed.